hl/src/H5LT.c

```diff
@@ -355,15 +355,28 @@ image_free(void *ptr, H5FD_file_image_op_t file_image_op, void *_udata)
 
             udata->vfd_ref_count--;
 
-            /* release the shared buffer only if indicated by the respective flag and there are no outstanding
-             * references */
-            if (udata->fapl_ref_count == 0 && udata->vfd_ref_count == 0 &&
-                !(udata->flags & H5LT_FILE_IMAGE_DONT_RELEASE)) {
-                free(udata->vfd_image_ptr);
-                udata->app_image_ptr  = NULL;
-                udata->fapl_image_ptr = NULL;
-                udata->vfd_image_ptr  = NULL;
-            } /* end if */
+            if (udata->fapl_ref_count == 0 && udata->vfd_ref_count == 0) {
+                /* release the shared buffer only if indicated by the respective
+                 * flag and there are no outstanding references
+                 */
+                if (!(udata->flags & H5LT_FILE_IMAGE_DONT_RELEASE)) {
+                    free(udata->vfd_image_ptr);
+                    udata->app_image_ptr  = NULL;
+                    udata->fapl_image_ptr = NULL;
+                    udata->vfd_image_ptr  = NULL;
```

Comment:
Should we assert here that these pointers point to the same location?

Reply:
I can add that assert if you think it's useful. I believe that depending on whether one actually opens the file image or just calls `H5Pset_file_image` to set a file image on a FAPL without ever opening a file (odd but could happen), then you could either have `udata->app_image_ptr == udata->fapl_image_ptr` and `udata->vfd_image_ptr == NULL` OR `udata->app_image_ptr == udata->fapl_image_ptr == udata->vfd_image_ptr`. In this particular block, I believe the latter should always be true, but I haven't verified that.

Reply:
I guess we should assert that everything is either the same or NULL, we don't want to leak more memory

Reply:
```
assert(udata->app_image_ptr == udata->vfd_image_ptr || udata->app_image_ptr == NULL);
assert(udata->fapl_image_ptr == udata->vfd_image_ptr || udata->fapl_image_ptr == NULL);
```